py_ecc/fields/optimized_field_elements.py

```diff
@@ -0,0 +1,395 @@
+                for c
+                in self.coeffs
+            ], self.curve_name)
+        elif isinstance(other, type(self)):
+            return self * other.inv()
+        else:
+            raise ValueError(
+                "Expected an int or FQP object, but got object of type {}"
+                .format(type(other))
+            )
+
+    def __truediv__(self, other: Union[int, "FQ", "FQP"]) -> "FQP":
+        return self.__div__(other)
+
+    def __pow__(self, other: int) -> "FQP":
+        o = type(self)([1] + [0] * (self.degree - 1), self.curve_name)
+        t = self
+        while other > 0:
+            if other & 1:
+                o = o * t
+            other >>= 1
+            t = t * t
+        return o
+
+    def optimized_poly_rounded_div(self,
+                                   a: Sequence[IntOrFQ],
+                                   b: Sequence[IntOrFQ]) -> Sequence[IntOrFQ]:
+        dega = deg(a)
+        degb = deg(b)
+        temp = [x for x in a]
+        o = [0 for x in a]
+        for i in range(dega - degb, -1, -1):
+            o[i] = int(o[i] + temp[degb + i] * prime_field_inv(int(b[degb]), self.field_modulus))
+            for c in range(degb + 1):
+                temp[c + i] = (temp[c + i] - o[c])
+        return [x % self.field_modulus for x in o[:deg(o) + 1]]
+
+    # Extended euclidean algorithm used to find the modular inverse
+    def inv(self) -> "FQP":
+        lm, hm = [1] + [0] * self.degree, [0] * (self.degree + 1)
+        low, high = (
+            cast(List[IntOrFQ], list(self.coeffs + (0,))),
+            cast(List[IntOrFQ], list(self.modulus_coeffs + (1,))),
+        )
+        low, high = list(self.coeffs + (0,)), self.modulus_coeffs + (1,)  # type: ignore
+        while deg(low):
+            r = cast(List[IntOrFQ], list(self.optimized_poly_rounded_div(high, low)))
+            r += [0] * (self.degree + 1 - len(r))
+            nm = [x for x in hm]
+            new = [x for x in high]
+            # assert len(lm) == len(hm) == len(low) == len(high) == len(nm) == len(new) == self.degree + 1  # noqa: E501
+            for i in range(self.degree + 1):
+                for j in range(self.degree + 1 - i):
+                    nm[i + j] -= lm[i] * int(r[j])
+                    new[i + j] -= low[i] * r[j]
+            nm = [x % self.field_modulus for x in nm]
+            new = [int(x) % self.field_modulus for x in new]
+            lm, low, hm, high = nm, new, lm, low
+        return type(self)(lm[:self.degree], self.curve_name) / low[0]
+
+    def __repr__(self) -> str:
+        return repr(self.coeffs)
+
+    def __eq__(self, other: "FQP") -> bool:     # type: ignore # https://github.com/python/mypy/issues/2783 # noqa: E501
+        if not isinstance(other, type(self)):
+            raise ValueError(
+                "Expected an FQP object, but got object of type {}"
+                .format(type(other))
+            )
+
+        for c1, c2 in zip(self.coeffs, other.coeffs):
```

Comment:
I think this had a bug in the original source, but seems out of scope. Opened #45